src/org/labkey/test/pages/core/login/AuthDialogBase.java

```diff
@@ -0,0 +1,112 @@
+package org.labkey.test.pages.core.login;
+
+import org.labkey.test.Locator;
+import org.labkey.test.components.bootstrap.ModalDialog;
+import org.labkey.test.components.html.Input;
+import org.labkey.test.components.html.ToggleButton;
+import org.labkey.test.params.login.AuthenticationProvider;
+import org.openqa.selenium.WebDriver;
+
+public abstract class AuthDialogBase<T extends AuthDialogBase<T>> extends ModalDialog
+{
+    private final LoginConfigRow _row;
+
+    protected AuthDialogBase(AuthenticationProvider<?> provider, WebDriver driver)
+    {
+        super(getFinder("Add New " + provider.getProviderName() + " Configuration", driver));
+        _row = null;
+    }
+
+    protected AuthDialogBase(LoginConfigRow row)
+    {
+        super(getFinder("Configure " + row.getDescription(), row.getDriver()));
+        _row = row;
+    }
+
+    protected AuthDialogBase(ModalDialogFinder finder)
+    {
+        super(finder);
+        _row = null;
+    }
+
+    private static ModalDialogFinder getFinder(String title, WebDriver driver)
+    {
+        return new ModalDialogFinder(driver).withTitle(title);
+    }
+
+    protected LoginConfigRow getRow()
+    {
+        return _row;
+    }
+
+    public T setDescription(String description)
+    {
+        elementCache().descriptionInput.set(description);
+        return getThis();
+    }
+
+    public String getDescription()
+    {
+        return elementCache().descriptionInput.get();
+    }
+
+    public T setEnabled(boolean enabled)
+    {
+        elementCache().enableToggle.set(enabled);
+        return getThis();
+    }
+
+    public boolean isEnabled()
+    {
+        return elementCache().enableToggle.get();
+    }
+
+    public T clickButtonExpectingError(String buttonText)
+    {
+        Locators.dismissButton(buttonText).waitForElement(this, 2000).click();
+        return getThis();
+    }
+
+    /**
+     * for UX reasons, the wording in the button to submit an auth dialog differs when we're creating
+     * a new config vs. editing an existing one.
+     */
+    public LoginConfigRow clickApply()
+    {
+        String description= getDescription();
+        Locator.findAnyElement("Finish or Apply button", this,
+                Locators.dismissButton("Finish"), Locators.dismissButton("Apply")).click();
+        waitForClose(4);
+        return new LoginConfigRow.LoginConfigRowFinder(getDriver()).withDescription(description).waitFor();
+    }
```

Comment:
We know whether this is a new config based on whether a `LoginConfigRow` was passed in. We could also use this knowledge to make `clickButtonExpectingError` not require button text.
```
dismiss(_row == null ? "finish" : "apply");
```

Reply:
The only downside to this is that it means you don't know the button's text by just reading test code.  So, readability downside, but maintainability win